### DISCUSSION TITLE
createEntityStore Api like Database Action

### DISCUSSION THREAD

**Discussion by cqh963852** [zerobias/effector#664] [Ideas]
I am looking for a eaiser api to create store like redux toolkit's createEntityAdapter.

```ts
type ID = string | number;
interface IData {
  id: ID;
}

export const $addOne = createEvent<IData>();
export const $removeOne = createEvent<ID>();

const $entityStore = createEntityStore<IData>()
  .on($addOne, (entityState, data) => {
    return entityState.addOne(data);
  })
  .on($removeOne, (entityState, id) => {
    return entityState.removeOne(id);
  });

export default $entityStore;
```

What about createEntityStore?

**Comment by cqh963852** [comment]
Another way.
I have fork redux-toolkit 's entity library

```ts
import { createEntityAdapter } from "@idealjs/entity";
import { createEvent, createStore } from "effector";

interface IUser {
  id: string;
  name: string;
}
const $addOne = createEvent<IUser>();
const $removeOne = createEvent<string>();

const entityAdapter = createEntityAdapter<IUser>();

const store = createStore(entityAdapter.getInitialState());

const $entityStore = store
  .on($addOne, (entityState, data) => {
    return entityAdapter.addOne(entityState, data);
  })
  .on($removeOne, (entityState, data) => {
    return entityAdapter.removeOne(entityState, data);
  });
```

It can also create entity data. 

But redux-toolkit's entity need immer,reselect as deps.

**Comment by zerobias** [comment]
I think this is a great idea. The only difficulty in implementing this idea is the need for continuous feedback from people who test the implementation in practice, on real-world tasks. So that you can ask: "Is this convenient? What if we do this? What is causing difficulties?"

Not so long ago [I tried to implement](https://codesandbox.io/embed/effector-todo-app-forked-d264v?codemirror=1) an api for working with collections (queries, filtering; most likely we will want to do more than just add and remove items, right?), but without feedback it was purely an abstract experiment. 

Notice how the target UI looks like — the key trick is that the editable and collapsible tree squeezes the most out of the approach

And how would you implement such a task? I found that a lot of issues comes in an interaction between api itself and the view

**Reply by cqh963852** [reply]
> queries, filtering; most likely we will want to do more than just add and remove items. 

I agree with this.

---- 
![image](https://user-images.githubusercontent.com/17702287/164204978-8a91b11c-d93c-4728-b6d3-ae52ec639359.png)
I have learn the todos app.

The createItemApi is great,It looks like redux-toolkit 's createSlice Api, but easier than it.

While using redux-toolkit's createSlice, I have to do reexport. just like

![image](https://user-images.githubusercontent.com/17702287/164206652-cfb94bb1-2f13-4bba-8bfd-46ce20c859e5.png)

`createItemApi ` doesn't need reexport,

----

But it is quite complex.

what about
```ts
  const [todo, api] = useItem(id, todoItemApi)
  // =>
  // const todo = useItem(id,todoItemApi) // select the todo entity
  // const api = useItemApi(id,todoItemApi) // select the todo entity's api
```
In this case, an interface is only responsible for one function

And some time, The view and the control are placed at different component.

**Reply by zerobias** [reply]
I updated the example, `useItem` was split into `useItemApi` and `useItemState` and the files `model` and `modelView` were merged
<img width="596" alt="image" src="https://user-images.githubusercontent.com/15912112/164392484-fe34a028-a7b7-4a67-8d5e-fde7f33dc9f9.png">

**Reply by zerobias** [reply]
Eventually, it would be good to see this api as an official part of effector, `effector/keyval` or `@effector/keyval`. But there is still the unsolved issue: how to initialize an item as a child of another one — I struggled to properly [get an ID](https://codesandbox.io/s/effector-todo-app-forked-d264v?from-embed=&file=/src/todo-app/model.ts) of the parent
![1BCE88F0-FD1D-421F-B3EA-97D7060250A4](https://user-images.githubusercontent.com/15912112/166270022-d9617e41-a63b-415f-9ade-57b367f3d315.png)
